lib/encodings/base64.fz

```diff
@@ -0,0 +1,155 @@
+      n1 => ((n >> 12) & 63).as_i32
+      n2 => ((n >> 6) & 63).as_i32
+      n3 => (n & 63).as_i32
+
+      # encode sextets
+      block := alphabet[n0] + alphabet[n1] + alphabet[n2] + alphabet[n3] # NYI: OPTIMIZATION: lots of string concatenation
+    else
+      res.substring 0 (res.byte_length - pad_size) + "=" * pad_size
+
+  # decodes a base64 string, decoding is strict as required by RFC 4648
+  # non alphabet characters, line breaks, missing padding cause errors
+  public decode_str(data String) =>
+    if data.byte_length != data.codepoint_length
+      # check for multi byte unicode chars
+      for c in data.as_codepoints
+          i in 0..
+          res :=
+            if (c.byte_length > 1)
+              error "character '$c' at position $i is not in $encoding_name alphabet"
+            else outcome true
+      until !res.ok
+        res.err
+      else error "this should never happen"
+
+    else
+      decode data.utf8.as_array
+
+
+  # decodes a sequence of ASCII characters, decoding is strict as required by RFC 4648
+  # non alphabet characters, line breaks, missing padding cause errors
+  public decode(data array u8) =>
+
+    # determine padding length '=' is 61 in ASCII
+    pad_size :=
+      if data.count > 1 && data[data.count - 1] = 61
+        if data[data.count - 2] = 61 then 2
+        else                              1
+      else                                0
+
+    alphabet_check :=
+      for b in data
+          i:=0, i+1
+      do
+        check_char :=
+          # base64 alphabet character
+          if (b >= 65 && b <= 90) || (b >= 97 && b <= 122) || (b >= 48 && b <= 57) || (b = char_62_ascii) || (b = get_char_63)
+            outcome true
+
+          # padding character =
+          else if b = 61
+            if i < data.length - pad_size
+              error """
+                    padding character '=' not allowed within the input data, \
+                    only at the very end, as required by RFC464"""
+            else outcome true
+
+          # linebreak
+          else if b = 10 || b = 13
+              error """
+                    line breaks are not allowed within encoded data, as required by RFC464, found \
+                    {if b=10 then "LF" else "CR"} at position $i"""
+          # other non alphabet character
+          else
+            error "byte $i in input data, representing '{codepoint b.as_u32}' in ASCII, is not in $encoding_name alphabet"
+
+      until !check_char.ok
+        check_char.err
+      else outcome true
+
+    if !alphabet_check.ok       # check for non base64 characters
+      alphabet_check.err
+
+    else if !(data.length %% 4) # check input length
+      error "length of input data is not multiple of four, as required by RFC4648"
+
+    else                        # do decoding
+
+      # replace padding with zeros (character 'A' / 65 is decoded to zero)
+      input := ((data.take(data.count - pad_size)) ++ ([(u8 65)].cycle.take pad_size)).as_array
```

Comment:
Can this be done with map `data.map x -> if x=61 then (u8 65) else x).as_array`?

Reply:
That should actually be possible now, since all the error handling is done before decoding the input